### PR TITLE
nydus: add lazyload support for kata with clh

### LINKED ISSUE #3654: nydus: support CLH hypervisor
The first [PR](https://github.com/kata-containers/kata-containers/pull/2977) of nydus only supports QEMU,
other hypervisors should be supported too.

/cc @luodw

### PR DESCRIPTION
**Pull request by luodw**
As kata with qemu has supported lazyload, so this pr aims to
bring lazyload ability to kata with clh.

Fixes #3654

Signed-off-by: luodaowen.backend <luodaowen.backend@bytedance.com>

### CODE REVIEW

File: src/runtime/virtcontainers/clh.go [hunk 1]
```diff
@@ -358,6 +362,25 @@ func (clh *cloudHypervisor) CreateVM(ctx context.Context, id string, network Net
 		cache:      clh.config.VirtioFSCache,
 	}
 
+	if clh.config.SharedFS == config.VirtioFSNydus {
+		apiSockPath, err := clh.nydusdAPISocketPath(clh.id)
+		if err != nil {
+			clh.Logger().WithError(err).Info("Invalid api socket path for nydusd")
```
Comment (jodh-intel) [inline src/runtime/virtcontainers/clh.go:368]: Nit:

```suggestion
			clh.Logger().WithError(err).Error("Invalid api socket path for nydusd")
```

**Review by jodh-intel — COMMENTED** [review]
Thanks @luodw!

Could you update https://github.com/kata-containers/kata-containers/blob/main/docs/how-to/how-to-use-virtio-fs-nydus-with-kata.md#how-to to either:

- State that "QEMU is being used as an example and other hypervisors may be supported - see issue ..."
- State that Nydus works with QEMU and CH.

lgtm

**Comment by liubin** [comment]
Hi @luodw I'm still seeiing the "4 workflows awaiting approval" button, seems that you are still a first-time contributor.

I checked the commit has been merged that showing:

```
From 2d9f89aec767de4b09cf365c74762aed2c201797 Mon Sep 17 00:00:00 2001
From: "luodaowen.backend" <luodaowen.backend@bytedance.com>
```

Had you connected this email to your [GitHub account emails setting](https://github.com/settings/emails)?

**Comment by luodw** [comment]
> Hi @luodw I'm still seeiing the "4 workflows awaiting approval" button, seems that you are still a first-time contributor.
> 
> I checked the commit has been merged that showing:
> 
> ```
> From 2d9f89aec767de4b09cf365c74762aed2c201797 Mon Sep 17 00:00:00 2001
> From: "luodaowen.backend" <luodaowen.backend@bytedance.com>
> ```
> 
> Had you connected this email to your [GitHub account emails setting](https://github.com/settings/emails)?

yes, `luodaowen.backend@bytedance.com` is my primary email, I can receive github notifications from this email.

File: src/runtime/virtcontainers/clh.go (hunk 1 above)
Comment (luodw) [reply src/runtime/virtcontainers/clh.go]: has done

**Comment by luodw** [comment]
> Thanks @luodw!
> 
> Could you update https://github.com/kata-containers/kata-containers/blob/main/docs/how-to/how-to-use-virtio-fs-nydus-with-kata.md#how-to to either:
> 
> * State that "QEMU is being used as an example and other hypervisors may be supported - see issue ..."
> * State that Nydus works with QEMU and CH.
> 
> lgtm

@jodh-intel have updated

**Review by jodh-intel — APPROVED** [review]
Thanks @luodw.

lgtm

File: docs/how-to/how-to-use-virtio-fs-nydus-with-kata.md [hunk 2]
```diff
@@ -16,15 +16,15 @@ You can use Kata Containers with `nydus` as follows,
 
 4. Use [kata-containers](https://github.com/kata-containers/kata-containers) `latest` branch to compile and build `kata-containers.img`;
 
-5. Update `configuration-qemu.toml` to include:
+5. Update `configuration-qemu.toml` or `configuration-qemu.toml`to include:
```
Comment (liubin) [inline docs/how-to/how-to-use-virtio-fs-nydus-with-kata.md:19]: or `configuration-qemu.toml` -> or `configuration-clh.toml` ?

File: docs/how-to/how-to-use-virtio-fs-nydus-with-kata.md [hunk 3]
```diff
@@ -16,15 +16,15 @@ You can use Kata Containers with `nydus` as follows,
...
 
 ```toml
 shared_fs = "virtio-fs-nydus"
 virtio_fs_daemon = "<nydusd binary path>"
 virtio_fs_extra_args = []
 ```
 
-6. run `crictl run -r kata-qemu nydus-container.yaml nydus-sandbox.yaml`;
+6. run `crictl run -r kata-qemu/kata-clh nydus-container.yaml nydus-sandbox.yaml`;
```
Comment (liubin) [inline docs/how-to/how-to-use-virtio-fs-nydus-with-kata.md:27]: I think `kata-qemu/kata-clh` can simply change to `kata`.

File: src/runtime/pkg/katautils/config.go [hunk 4]
```diff
@@ -426,7 +426,7 @@ func (h hypervisor) sharedFS() (string, error) {
 	supportedSharedFS := []string{config.Virtio9P, config.VirtioFS, config.VirtioFSNydus}
 
 	if h.SharedFS == "" {
-		return config.Virtio9P, nil
+		return config.VirtioFS, nil
```
Comment (liubin) [inline src/runtime/pkg/katautils/config.go:429]: Why changes to virtiofs? This is a big change that is worth writing to changelog/release notes.

Comment (luodw) [reply src/runtime/pkg/katautils/config.go:429]: I remember kata has set default  to `virtiofsd` before? I think it need fix so I updated.

File: src/runtime/pkg/katautils/config.go [hunk 5]
```diff
@@ -822,7 +822,10 @@ func newClhHypervisorConfig(h hypervisor) (vc.HypervisorConfig, error) {
 		return vc.HypervisorConfig{}, err
 	}
 
-	sharedFS := config.VirtioFS
+	sharedFS, err := h.sharedFS()
```
Comment (liubin) [inline src/runtime/pkg/katautils/config.go:825]: Here are some virtiofs/nydusd checks for qemu:  
https://github.com/kata-containers/kata-containers/blob/cf53ec2c714d3dd665806c07d92120ae869f1573/src/runtime/pkg/katautils/config.go#L647-L655 
CLH also needs them.

File: src/runtime/pkg/katautils/config.go (hunk 4 above)
Comment (liubin) [reply src/runtime/pkg/katautils/config.go:429]: I'm sorry for that, truely the default is virtiofs: https://github.com/kata-containers/kata-containers/blob/be2e90469a48287091161d80e01fe473c110d7e7/src/runtime/Makefile#L166

Please ignore this.

File: docs/how-to/how-to-use-virtio-fs-nydus-with-kata.md (hunk 3 above)
Comment (luodw) [reply docs/how-to/how-to-use-virtio-fs-nydus-with-kata.md]: agree, have updated

File: src/runtime/pkg/katautils/config.go (hunk 5 above)
Comment (luodw) [reply src/runtime/pkg/katautils/config.go:820]: have updated

**Review by jodh-intel — APPROVED** [review]
Thanks @luodw.

lgtm

**Comment by liubin** [comment]
@luodw still an unit test failed:

> === RUN   TestNewQemuHypervisorConfig
    config_test.go:658: cannot enable virtio-fs without daemon path in configuration file
--- FAIL: TestNewQemuHypervisorConfig (0.00s)

**Comment by luodw** [comment]
> @luodw still an unit test failed:
> 
> > === RUN   TestNewQemuHypervisorConfig
> > config_test.go:658: cannot enable virtio-fs without daemon path in configuration file
> > --- FAIL: TestNewQemuHypervisorConfig (0.00s)

have udpated, I think it is ok now.

**Review by liubin — APPROVED** [review]

**Review by jodh-intel — APPROVED** [review]
Thanks @luodw.

lgtm

**Pull request merged**